File: 40. IsomorphicStrings.java
public class IsomorphicStrings {

    public static boolean isIsomorphic(String s, String t) {
        int[] ms = new int[256], mt = new int[256];
        for (int i = 0; i < s.length(); i++) {
            char a = s.charAt(i), b = t.charAt(i);
            if (ms[a] != mt[b]) return false;
            ms[a] = mt[b] = i + 1;
        }
        return true;
    }

    public static void main(String[] args) {
        System.out.println(isIsomorphic("egg", "add"));     // true
        System.out.println(isIsomorphic("foo", "bar"));     // false
        System.out.println(isIsomorphic("paper", "title")); // true
    }
}

File: 40. IsomorphicStrings.c
#include <stdbool.h>
#include <stdio.h>

bool isIsomorphic(const char *s, const char *t) {
    int ms[256] = {0}, mt[256] = {0};
    for (int i = 0; s[i] && t[i]; i++) {
        unsigned char a = (unsigned char)s[i], b = (unsigned char)t[i];
        if (ms[a] != mt[b]) return false;
        ms[a] = mt[b] = i + 1;
    }
    return true;
}

int main(void) {
    puts(isIsomorphic("egg",   "add")   ? "true" : "false"); // true
    puts(isIsomorphic("foo",   "bar")   ? "true" : "false"); // false
    puts(isIsomorphic("paper", "title") ? "true" : "false"); // true
    return 0;
}
